tests/integration/test_sns.py

```diff
@@ -2330,3 +2328,172 @@ def validate_content():
             sleep_before = 10
 
         retry(validate_content, retries=retries, sleep_before=sleep_before, sleep=sleep)
+
+    @pytest.mark.only_localstack
+    def test_publish_to_platform_endpoint_can_retrospect(
+        self, sns_client, sns_create_topic, sns_subscription
+    ):
+        sns_backend = SNSBackend.get()
+        # clean up the saved messages
+        sns_backend_endpoint_arns = list(sns_backend.platform_endpoint_messages.keys())
+        for saved_endpoint_arn in sns_backend_endpoint_arns:
+            sns_backend.platform_endpoint_messages.pop(saved_endpoint_arn, None)
+
+        topic_arn = sns_create_topic()["TopicArn"]
+        application_platform_name = f"app-platform-{short_uid()}"
+
+        app_arn = sns_client.create_platform_application(
+            Name=application_platform_name, Platform="p1", Attributes={}
+        )["PlatformApplicationArn"]
+
+        endpoint_arn = sns_client.create_platform_endpoint(
+            PlatformApplicationArn=app_arn, Token=short_uid()
+        )["EndpointArn"]
+
+        endpoint_arn_2 = sns_client.create_platform_endpoint(
+            PlatformApplicationArn=app_arn, Token=short_uid()
+        )["EndpointArn"]
+
+        sns_subscription(
+            TopicArn=topic_arn,
+            Protocol="application",
+            Endpoint=endpoint_arn,
+        )
+
+        # example message from
+        # https://docs.aws.amazon.com/sns/latest/dg/sns-send-custom-platform-specific-payloads-mobile-devices.html
+        message = json.dumps({"APNS_PLATFORM": json.dumps({"aps": {"content-available": 1}})})
+        message_for_topic = json.dumps(
+            {
+                "default": "This is the default message which must be present when publishing a message to a topic.",
+                "APNS_PLATFORM": json.dumps({"aps": {"content-available": 1}}),
+            },
+        )
+        message_attributes = {
+            "AWS.SNS.MOBILE.APNS.TOPIC": {
+                "DataType": "String",
+                "StringValue": "com.amazon.mobile.messaging.myapp",
+            },
+            "AWS.SNS.MOBILE.APNS.PUSH_TYPE": {
+                "DataType": "String",
+                "StringValue": "background",
+            },
+            "AWS.SNS.MOBILE.APNS.PRIORITY": {
+                "DataType": "String",
+                "StringValue": "5",
+            },
+        }
+        # publish to a topic which has a platform subscribed to it
+        sns_client.publish(
+            TopicArn=topic_arn,
+            Message=message_for_topic,
+            MessageAttributes=message_attributes,
+            MessageStructure="json",
+        )
+        # publish directly to the platform endpoint
+        sns_client.publish(
+            TargetArn=endpoint_arn_2,
+            Message=message,
+            MessageAttributes=message_attributes,
+            MessageStructure="json",
+        )
+
```

Comment:
Nit: It would be nice if we had at least one instance where we don't set 'MessageStructure', just to have that code covered somewhere. But that is probably not the overall purpose of the test.

Reply:
You're right. I'm having issue as AWS examples in the docs always set the MessageStructure to `json`,  and I have no way for now to test it against AWS. You need credentials to create a platform application, and we would need to create one either in Firebase Cloud Messaging, or Apple, or any other. We will need to get to it at some point to validate behaviours and format. But I really agree with you that this is best case scenario, and we don't really have any idea of what is really happening in case of errors.